### ISSUE TITLE
rule: `always-set-diagnostic-before-error`

### ISSUE THREAD

**Issue by MichaelBelousov** [DonIsaac/zlint#332]
## Description
Make sure that code following (a slightly modified version of) the [diagnostic pattern](https://github.com/ziglang/zig/issues/2647#issuecomment-589829306) set their diagnostic to a necessary data right before returning any errors.

I think that the diagnostic pattern is too flexible considering the diagnostic could be an argument, a field of an options argument, might not be a tagged union, may require separate initialization for some fields, etc. But I wanted to discuss, since zlint might be a good engine to prevent these bugs for me. I am also not against contributing the necessary stuff myself once I have some time for it.

One (probably way too intense) thing that would be good for my particular situation is like a doc-comment to indicate an arg or field that should always be "set" before returning an error. 

## Examples

Examples of **incorrect** code for this rule
```zig
fn parse(src: []const u8, diag: ?*Diag) !Data {
  if (startsWith(u8, "hello", src)) {
    return error.NeverSayHello;
  }
  //...
  return Data{};
}
```

Examples of **correct** code for this rule
```zig
fn parse(src: []const u8, in_diag: ?*Diag) !Data {
  var dummy_diag: Diag = undefined;
  const diag = in_diag orelse dummy_diag;

  if (startsWith(u8, "hello", src)) {
    diag.* = .{ .NeverSayHello = relevant_data };
    return error.NeverSayHello;
  }
  //...
  return Data{};
}
```

**Comment by MichaelBelousov** [comment]
I will withdraw this and maybe implement it as a custom rule once #411 is merged

**Comment by MichaelBelousov** [comment]
(because it is nuanced)
